inc/Engine/Optimization/DeferJS/DeferJS.php

```diff
@@ -75,6 +75,83 @@ public function defer_js( string $html ) : string {
 		return $html;
 	}
 
+	/**
+	 * Defers inline JS containing jQuery calls
+	 *
+	 * @since 3.8
+	 *
+	 * @param string $html HTML content.
+	 * @return string
+	 */
+	public function defer_inline_js( string $html ) : string {
+		if ( ! $this->can_defer_js() ) {
+			return $html;
+		}
+
+		$exclude_defer_js = implode( '|', $this->get_excluded() );
+
+		if ( preg_match( '/(jquery(?:.*)?\.js)/i', $exclude_defer_js ) ) {
+			return $html;
+		}
+
+		$buffer_nocomments = preg_replace( '/<!--(.*)-->/Uis', '', $html );
+
+		preg_match_all( '#<script(?:[^>]*)>(?<content>[\s\S]*?)</script>#msi', $buffer_nocomments, $matches, PREG_SET_ORDER );
+
+		if ( empty( $matches ) ) {
+			return $html;
+		}
+
+		/**
+		 * Filters the patterns used to find jQuery in inline JS
+		 *
+		 * @since 3.8
+		 *
+		 * @param string $jquery_patterns A RegEx pattern to find jQuery inline JS.
+		 */
+		$jquery_patterns = apply_filters( 'rocket_defer_jquery_patterns', 'jQuery|\$\.\(|\$\(' );
+
+		/**
+		 * Filters the patterns used to find inline JS that should not be deferred
+		 *
+		 * @since 3.8
+		 *
+		 * @param string $inline_exclusions A RegEx pattern to find inline JS that should not be deferred.
+		 */
+		$inline_exclusions = apply_filters( 'rocket_defer_inline_exclusions', 'DOMContentLoaded|document\.write' );
+
+		foreach ( $matches as $inline_js ) {
+			if ( empty( $inline_js['content'] ) ) {
+				continue;
+			}
+
+			if ( preg_match( "/({$inline_exclusions})/msi", $inline_js['content'] ) ) {
+				continue;
+			}
+
+			if ( ! preg_match( "/({$jquery_patterns})/msi", $inline_js['content'] ) ) {
+				continue;
+			}
+
+			$tag  = str_replace( $inline_js['content'], $this->inline_js_wrapper( $inline_js['content'] ), $inline_js[0] );
+			$html = str_replace( $inline_js[0], $tag, $html );
```

Comment:
I was talking with @hellofromtonya last night about this:-
which approach would be better (performance wise)

Approach 1
```
LOOP START
str_replace(string1, string2, subject);
LOOP END
```
Approach 2
```
LOOP START
array[string1] = string2;
LOOP END
str_replace(array_keys(array), array, subject);
```

I think Approach 2 would be faster, but I added this as a hint if you have input ( really it's for me to know for my knowledge ).

Reply:
Our conversation was hypothetical. Now that I see the implementation in its context, this line of code bothers me

```
$html = str_replace( $inline_js[0], $tag, $html );
```

Why? It means the entire HTML string is walked to search and then replace matches on each iteration of the `foreach`. 

Let me think about this and do some benchmarking to determine which is more performant.

Reply:
Benchmarked 2 approaches

- Approach 1: code above [link](https://3v4l.org/jXYuV)
- Approach 2: do the HTML search once [link](https://3v4l.org/aVprY)
```php
$replace = [];
$search  = [];
        
foreach ( $matches as $inline_js ) {
	if ( empty( $inline_js['content'] ) ) {
		continue;
	}

	if ( preg_match( "/({$inline_exclusions})/msi", $inline_js['content'] ) ) {
		continue;
	}

	if ( ! preg_match( "/({$jquery_patterns})/msi", $inline_js['content'] ) ) {
		continue;
	}

	$search[]   = $inline_js[0];
	$replace[]  = str_replace( $inline_js['content'], $this->inline_js_wrapper( $inline_js['content'] ), $inline_js[0] );
		}
		
	$html = str_replace( $search, $replace, $html );
````

Notes:

- Used `microtime` as a cheap benchmark
- 10,000 cycles for an average time for each PHP version
- [Here's a link to the spreadsheet](https://docs.google.com/spreadsheets/d/1Njo2A_OS4-ekAUP9pFlLf0m4IqOBd-jeRCdFKq3i-1k/edit?usp=sharing)

Results:

|                                        | Approach 1 | Approach 2 |
|----------------------- | ----------- | ------------ |
| Average microseconds |           8.62  |             9.69  |
|Total number of cycles  |     620,000 |       640,000 |
| code                               | [link](https://3v4l.org/jXYuV) | [link](https://3v4l.org/aVprY) |


Conclusion:
Go with the code in this PR. Why?

1. Teeny tiny bit faster 
2. Uses less memory
3. Less code
4. Easier to understand
5. Easier to maintain